### PR TITLE
fix: improve typings for the filter operator

### LINKED ISSUE #17: Improve typing for "filter"
Currently `filter` assumes the same return value, even if the passed predicate is a type predicate (with the return type with `is`). Here's how typings for ES5 solved this:

```ts
filter<S extends T>(callbackfn: (value: T, index: number, array: T[]) => value is S, thisArg?: any): S[];
filter(callbackfn: (value: T, index: number, array: T[]) => any, thisArg?: any): T[];
```

Would be nice to see the same in iterare.

### PR DESCRIPTION
**Pull request by lazarljubenovic**
closes #17

Do you have a better idea bout how to test type information? It's pretty hard to Google too since I keep getting results about how to setup testing frameworks in TS instead of how to assert for types. I guess we'd need to use the compiler API?

### CODE REVIEW

**Comment by felixfbecker** [comment]
In general I think typings like this are not tested. But you could do it statically by just assigning the result to a variable with a type declaration, i.e.
```ts
const num: number = [1, 'a'].filter((x: string | number): x is number => typeof x === 'number')
```
Would fail to compile if the overload wasn't there.

**Comment by lazarljubenovic** [comment]
Indeed, no need for the whole ceremony with a random `number` method. Will be patching it up later today;  thanks.

**Comment by lazarljubenovic** [comment]
Hmm... But given the low-level-ness of the tests since they're unit and not integration I guess makes it awkward to do what you suggest, since the result is the generic `FilterIterator<string | number, number>`. Maybe I should remove the test completely... :thinking:

**Comment by lazarljubenovic** [comment]
Hmm, there's this: https://github.com/dsherret/conditional-type-checks Could try it out later.

**Comment by felixfbecker** [comment]
If anything I would use https://github.com/Microsoft/dtslint#write-tests but I think it's overkill for this library. I would be fine to merge this without the test.

**Comment by lazarljubenovic** [comment]
Oh, didn't know there was an official thing as well. Alright then, removing the test later.

**Comment by felixfbecker** [comment]
Wanna remove the test?

**Comment by cojack** [comment]
@felixfbecker u gonne to accept it?

File: src/filter.test.ts [hunk 1]
```diff
@@ -10,4 +10,10 @@ describe('FilterIterator', () => {
         assert.equal(filtered.next().value, 6)
         assert.equal(filtered.next().done, true)
     })
+    it('should correctly infer type of the resulting iterable', () => {
+        const collection: IterableIterator<string | number> = [1, 'a'][Symbol.iterator]()
+        const isNumber = (x: string | number): x is number => typeof x === 'number'
+        const filtered = new FilterIterator(collection, isNumber)
+        assert.equal(filtered.next().value.toFixed(2), '1.00') // tsc will fail if .value is not of type "number"
+    })
```
Comment (felixfbecker) [inline src/filter.test.ts]:
```suggestion
```

**Review by felixfbecker — APPROVED** [review]

**Comment by lazarljubenovic** [comment]
Forgot about it and then it got lost in other notifications. Sorry and thanks for finishing the dirty work. :smile:

**Pull request merged**